server/src/main/java/org/elasticsearch/action/admin/indices/refresh/RefreshResponse.java

```diff
@@ -21,18 +21,59 @@
 
 import org.elasticsearch.action.ShardOperationFailedException;
 import org.elasticsearch.action.support.broadcast.BroadcastResponse;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
 
+import java.io.IOException;
 import java.util.List;
 
+import static org.elasticsearch.common.xcontent.ConstructingObjectParser.constructorArg;
+
 /**
  * The response of a refresh action.
  */
-public class RefreshResponse extends BroadcastResponse {
+public class RefreshResponse extends BroadcastResponse implements ToXContentFragment {
+
+    private static final ConstructingObjectParser<RefreshResponse, Void> PARSER = new ConstructingObjectParser<>("refresh",
+        true, arg -> (RefreshResponse) arg[0]);
+
+    static {
+        ConstructingObjectParser<RefreshResponse, Void> shardsParser = new ConstructingObjectParser<>("_shards", true,
+            arg -> new RefreshResponse((int) arg[0], (int) arg[1], (int) arg[2], null));
+        shardsParser.declareInt(constructorArg(), new ParseField(Fields.TOTAL));
+        shardsParser.declareInt(constructorArg(), new ParseField(Fields.SUCCESSFUL));
+        shardsParser.declareInt(constructorArg(), new ParseField(Fields.FAILED));
+        PARSER.declareObject(constructorArg(), shardsParser, new ParseField(Fields._SHARDS));
```

Comment:
I think that we never print out nor parse back the shard failures. Should we rather do so? Looking at the current behaviour, I think that for this API we return an http error code (taken from the first shard failure) , yet we may not want to throw an exception in such but rather return a proper `RefreshResponse` that includes shard failures? Not sure though, especially because I don't know what error codes we would have to set to be ignored in `RestHighLevelClient`. Maybe we would already have the exception with shard failures returned in this case. @tlrx what do you think?

Reply:
> I think that we never print out nor parse back the shard failures. Should we rather do so?

We do it (kind of) in `ReplicationResponse`: it contains `ShardInfo` that have failures which are `ShardOperationFailedException`. I'm not sure that we can reuse the code easily but I don't think that we should ignore the shard failures and always return `null`.

I also don't like the fact that `BroadcastResponse` picks up the first shard failure for its HTTP response code, maybe there's a rationale around this but I don't know it. Since the error codes could be anything, I'm afraid we'll need to parse the very first fields before being able to know if it's an exception or a refresh response with failures.

Reply:
I've a quick look at what we have done in `ReplicationResponse`'s `ShardInfo`, so here we shoud also do the same thing to parse and transfer shard failures ?

Reply:
yea we should do something similar. To make this feasible though we need #28312 to go in, so that it's enforced that only one type of exception needs to be parsed (`ShardOperationFailedException` is an interface)